digitaltwin_client/samples/digitaltwin_sample_sdk_info/digitaltwin_sample_sdk_info.c

```diff
@@ -0,0 +1,194 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <limits.h>
+
+#include "digitaltwin_client_version.h"
+#include "digitaltwin_sample_sdk_info.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_c_shared_utility/crt_abstractions.h"
+
+// DigitalTwin interface id SDK Information interface ID
+static const char DigitalTwinSampleSdkInfo_InterfaceId[] = "urn:azureiot:Client:SDKInformation:1";
+
+// DigitalTwin component name from service perspective.
+static const char DigitalTwinSampleSdkInfo_ComponentName[] = "urn_azureiot_Client_SDKInformation";
+
+//
+//  Property names and data for DigitalTwin read-only properties for this interface.
+//
+static const char DT_SdkLanguage_Property[] = "language";
+static const char DT_SdkVersion_Property[] = "version";
+static const char DT_SdkVendor_Property[] = "vendor";
+static const char DT_SdkLanguage[] = "C";
+static const char DT_SdkVendor[] = "Microsoft";
+
+//
+// Application state associated with the particular interface.  It contains 
+// the DIGITALTWIN_INTERFACE_CLIENT_HANDLE used for responses reporting properties.
+//
+typedef struct DIGITALTWIN_SAMPLE_SDK_INFO_TAG
+{
+    DIGITALTWIN_INTERFACE_CLIENT_HANDLE interfaceClientHandle;
+} DIGITALTWIN_SAMPLE_SDK_INFO_STATE;
+
+
+// State for interface.  For simplicity we set this as a global and set during DigitalTwin_InterfaceClient_Create, but  
+// callbacks of this interface don't reference it directly but instead use userContextCallback passed to them.
+static DIGITALTWIN_SAMPLE_SDK_INFO_STATE digitaltwinSample_SdkInfoState;
+
+
+// DigitalTwinSampleSdkInfo_PropertyCallback is invoked when a property is updated (or failed) going to server.
+// In this sample, we route ALL property callbacks to this function and just have the userContextCallback set
+// to the propertyName.  Product code will potentially have context stored in this userContextCallback.
+static void DigitalTwinSampleSdkInfo_PropertyCallback(DIGITALTWIN_CLIENT_RESULT dtReportedStatus, void* userContextCallback)
+{
+    if (dtReportedStatus == DIGITALTWIN_CLIENT_OK)
+    {
+        LogInfo("SDK_INFO: Property callback property=<%s> succeeds", (const char*)userContextCallback);
+    }
+    else
+    {
+        LogError("SDK_INFO: Property callback property=<%s> fails, error=<%s>", (const char*)userContextCallback, MU_ENUM_TO_STRING(DIGITALTWIN_CLIENT_RESULT, dtReportedStatus));
+    }
+}
+
+
+//
+// DigitalTwinSampleSdkInfo_ReportPropertyAsync is a helper function to report a SdkInfo's properties.
+// It invokes underlying DigitalTwin API for reporting properties and sets up its callback on completion.
+//
+static DIGITALTWIN_CLIENT_RESULT DigitalTwinSampleSdkInfo_ReportPropertyAsync(DIGITALTWIN_INTERFACE_CLIENT_HANDLE interfaceHandle, const char* propertyName, const char* propertyData)
+{
+    char* serializedPropertyValue = NULL;
+    DIGITALTWIN_CLIENT_RESULT result = DIGITALTWIN_CLIENT_ERROR;
+
+    size_t propertyDataSize = snprintf(NULL, 0, "\"%s\"", propertyData);
```

Comment:
>size_t propertyDataSize = snprintf(NULL, 0, "\"%s\"", propertyData); [](start = 4, length = 68)

I think we can remove the snprintf() and calloc().  What if instead we had DT_SdkLanguage and DT_SdkVendor just hard-code in the escaping, with a comment about the fact it had to be a legal jvalue.  For DigitalTwin_Client_GetVersionString() we could also use implicit string concat.  So something like DT_Version = "\"" DIGITALTWIN_CLIENT_SDK_VERSION "\"".

Reply:
I'm signing off momentarily, assuming you're in agreement with this comment.  If I'm missing something please let's talk when you're in prior to checkin.

---
In reply to: [341720075](https://github.com/Azure/azure-iot-sdk-c/pull/1289#discussion_r341720075) [](ancestors = 341720075)